### PR TITLE
Swerve and better PID classes

### PR DESCRIPTION
**Pull request by laviRZ**

### CODE REVIEW

File: src/main/java/frc/robot/components/TrigonXboxController.java [hunk 1]
```diff
@@ -106,13 +111,31 @@ public void intermittentRumble(int quantity) {
...
+            return 0;
+        return squared ? Math.signedSquare(super.getRightX()) : super.getRightX();
+    }
+
+    @Override
+    public double getRightY() {
+        if(super.getRightY() < deadband && super.getRightY() > -deadband)
+            return 0;
+        return squared ? Math.signedSquare(super.getRightY()) : super.getRightY();
     }
```
Comment (Mboesn) [inline src/main/java/frc/robot/components/TrigonXboxController.java:139]: make a deadband function and a squared function and have these functions returned something like 
`processedValue(super.getRightY())`

File: src/main/java/frc/robot/constants/RobotComponents.java [hunk 2]
```diff
@@ -17,15 +18,16 @@
     }
 
     protected static class SwerveComponents {
-        public static final Pigeon PIGEON = new Pigeon(CAN.Swerve.PIGEON_ID);
+        public static final Pigeon PIGEON = new Pigeon(new TrigonTalonSRX(12));
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/RobotComponents.java:21]: have it get a motor from a different subsystem and if it doesn't have it be
` new TrigonTalonSRX(CAN.Swerve.PIGEON_ID) `

File: src/main/java/frc/robot/constants/RobotComponents.java [hunk 3]
```diff
@@ -47,52 +49,73 @@
...
+                            LOCAL_SWERVE_MODULES_CONSTANTS.rearRightModuleConstants.angleCoefs).
+                    withRemoteSensorSource(
+                            CAN.Swerve.REAR_RIGHT_ANGLE_ENCODER_ID, RemoteSensorSource.TalonSRX_SelectedSensor, 0).
+                    withSecondaryFeedbackDevice(FeedbackDevice.RemoteSensor0),
+                    ControlMode.Position);
+            public static final PIDFTalonFX DRIVE_MOTOR = new PIDFTalonFX(
+                    CAN.Swerve.REAR_RIGHT_DRIVE_MOTOR_ID, DRIVE_MOTOR_CONFIG.withPID(
+                    LOCAL_SWERVE_MODULES_CONSTANTS.rearRightModuleConstants.driveCoefs),
+                    ControlMode.Velocity);
+        }
```
Comment (Mboesn) [inline src/main/java/frc/robot/constants/RobotComponents.java:119]: 👍

File: src/main/java/frc/robot/subsystems/swerve/SwerveModule.java [hunk 4]
```diff
@@ -96,7 +100,9 @@ private static double placeInAppropriate0To360Scope(
         };
     }
 
-    public void setDesiredState(SwerveModuleState desiredState, boolean isOpenLoop) {
+    public void setDesiredState(SwerveModuleState desiredState, boolean isOpenLoop, boolean isTuning) {
+        if(!isTuning && isTuning())
+            return;
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/swerve/SwerveModule.java:105]: explain why?

File: src/main/java/frc/robot/subsystems/swerve/SwerveSS.java [hunk 5]
```diff
@@ -45,6 +46,8 @@ public SwerveSS() {
      *                      each module's speed
      */
     public void drive(Translation2d translation, double rotation, boolean fieldRelative, boolean isOpenLoop) {
+        translation = new Translation2d(translation.getY(), -translation.getX());
+        rotation *= -1;
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/swerve/SwerveSS.java:50]: please write documentation as to why this is. to me it seems this might be solved by changing the motor ids (it thinks the side is the front)

File: src/main/java/frc/robot/utilities/LinearRegression.java [hunk 6]
```diff
@@ -1,5 +1,7 @@
 package frc.robot.utilities;
 
+import java.lang.Math;
+
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/LinearRegression.java:4]: why did you add an import and not do anything?

File: src/main/java/frc/robot/utilities/Math.java [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package frc.robot.utilities;
+
+public class Math {
+    public static double signedSquare(double x) {
+        return x * x * java.lang.Math.signum(x);
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/Math.java:5]: please just import Math

File: src/main/java/frc/robot/utilities/MotorConfig.java [hunk 8]
```diff
@@ -19,9 +19,12 @@
...
     private SupplyCurrentLimitConfiguration currentLimitConfig;
-    private FeedbackDevice feedbackDevice;
-    private RemoteSensorSource remoteSensorSourceType;
-    private int remoteSensorSourceDeviceId;
+    private FeedbackDevice primaryFeedbackDevice;
+    private FeedbackDevice secondaryFeedbackDevice;
+    private RemoteSensorSource remoteSensorSource0Type;
+    private RemoteSensorSource remoteSensorSource1Type;
+    private int remoteSensorSource0DeviceId;
+    private int remoteSensorSource1DeviceId;
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/MotorConfig.java:27]: why would we be using 2?

File: src/main/java/frc/robot/utilities/pid/PIDFMotor.java [hunk 9]
```diff
@@ -1,12 +1,22 @@
...
+import edu.wpi.first.util.sendable.SendableBuilder;
 
 /**
  * An interface used for motors that have a configurable PID
  */
 public interface PIDFMotor extends PIDFConfigurable {
     void setCoefs(PIDFCoefs pidfCoefs);
 
-    void tunePIDF(ControlMode controlMode);
+    void set(double output);
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/pid/PIDFMotor.java:11]: why can we remove tunePIDF exactly?

File: src/main/java/frc/robot/utilities/pid/PIDFTalon.java [hunk 10]
```diff
@@ -21,20 +20,20 @@ default void setCoefs(PIDFCoefs pidfCoefs) {
...
 
-    default void tunePIDF(ControlMode controlMode) {
-        set(controlMode, getTuningSetpoint(), DemandType.ArbitraryFeedForward, getCoefs().getKS());
-    }
-
     double getTuningSetpoint();
 
     default void setWithF(ControlMode controlMode, double setpoint) {
+        if(isTuning() && setpoint != getTuningSetpoint())
+            return;
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/pid/PIDFTalon.java:27]: document this weird failsafe

File: src/main/java/frc/robot/utilities/pid/PIDFTalonFX.java [hunk 11]
```diff
@@ -43,12 +52,18 @@ public double getTuningSetpoint() {
     }
 
     @Override
-    public void setSetpoint(double setpoint) {
+    public void setTuningSetpoint(double setpoint) {
         tuningSetpoint = setpoint;
+        setWithF(controlMode, setpoint);
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/pid/PIDFTalonFX.java:57]: why would setting the setpoint also move the motor?

**Review by Mboesn — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/swerve/SwerveSS.java (hunk 5 above)
Comment (laviRZ) [reply src/main/java/frc/robot/subsystems/swerve/SwerveSS.java:50]: no, it's because WPILib works with positive y is to the left, and positive x is forwards, and CCW+.

File: src/main/java/frc/robot/utilities/MotorConfig.java (hunk 8 above)
Comment (laviRZ) [reply src/main/java/frc/robot/utilities/MotorConfig.java:27]: if we want the angle motor to have its integrated sensor _and_ the mag encoder, we need to use two feedback devices

File: src/main/java/frc/robot/utilities/pid/PIDFMotor.java (hunk 9 above)
Comment (laviRZ) [reply src/main/java/frc/robot/utilities/pid/PIDFMotor.java:11]: because we have `setSetpoint()` defined in `PIDConfigurable`

File: src/main/java/frc/robot/constants/RobotComponents.java (hunk 3 above)
Comment (laviRZ) [reply src/main/java/frc/robot/constants/RobotComponents.java:120]: what?

File: src/main/java/frc/robot/utilities/pid/PIDFTalonFX.java (hunk 11 above)
Comment (laviRZ) [reply src/main/java/frc/robot/utilities/pid/PIDFTalonFX.java]: because the setpoint is for the motor...

File: src/main/java/frc/robot/RobotContainer.java [hunk 12]
```diff
@@ -34,19 +39,20 @@ public RobotContainer() {
...
         driveWithXboxCMD = new SupplierDriveCMD(
                 swerveSS,
-                squared ? () -> Math.pow(driverXbox.getLeftX(), 2) : driverXbox::getLeftX,
-                squared ? () -> Math.pow(driverXbox.getLeftY(), 2) : driverXbox::getLeftY,
-                squared ? () -> Math.pow(driverXbox.getRightX(), 2) : driverXbox::getRightX,
+                driverXbox::getLeftX,
+                driverXbox::getLeftY,
+                driverXbox::getRightX,
                 true);
+        SmartDashboard.putData("Drive CMD", driveWithXboxCMD);
```
Comment (Mboesn) [inline src/main/java/frc/robot/RobotContainer.java:55]: remove

File: src/main/java/frc/robot/RobotContainer.java [hunk 13]
```diff
@@ -34,19 +39,20 @@ public RobotContainer() {
      */
     private void initializeSubsystems() {
         swerveSS = new SwerveSS();
+        SmartDashboard.putData("Swerve", swerveSS);
```
[the remark this answers is not in the document]
Comment (Mboesn) [reply src/main/java/frc/robot/RobotContainer.java:42]: it might be but it's gonna get cluttered very quickly

File: src/main/java/frc/robot/components/TrigonXboxController.java [hunk 14]
```diff
@@ -127,4 +141,14 @@ public void notifierPeriodic() {
             rumbleAmount--;
         }
     }
+
+    private static double signedSquare(double value) {
+        return Math.signum(value) * Math.pow(Math.abs(value), 2);
+    }
+
+    private double processValue(double value) {
```
Comment (Mboesn) [inline src/main/java/frc/robot/components/TrigonXboxController.java:149]: documentation

File: src/main/java/frc/robot/subsystems/swerve/SwerveModule.java [hunk 15]
```diff
@@ -183,4 +207,53 @@ public SwerveModuleState getState() {
...
+        builder.addDoubleProperty(
+                "Angle Error",
+                () -> EncoderConversions.falconToDegrees(
+                        angleMotor.getClosedLoopError(), SwerveConstants.ANGLE_GEAR_RATIO)
+                , null);
+        builder.addBooleanProperty(
+                "Is Tuning",
+                this::isTuning
+                , this::setTuning);
+    }
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/swerve/SwerveModule.java:250]: put this below isTuning and setTuning

File: src/main/java/frc/robot/subsystems/swerve/SwerveSS.java (hunk 5 above)
Comment (Mboesn) [reply src/main/java/frc/robot/subsystems/swerve/SwerveSS.java:50]: 1. why hasn't this been an issue up to now
2. add documentation

File: src/main/java/frc/robot/utilities/pid/PIDFMotor.java (hunk 9 above)
Comment (Mboesn) [reply src/main/java/frc/robot/utilities/pid/PIDFMotor.java:11]: setSetpoint does not actually move the motor. in the pidController you need to use caculate and in talons you need to do set

File: src/main/java/frc/robot/utilities/pid/PIDFTalon.java [hunk 16]
```diff
@@ -21,20 +20,27 @@ default void setCoefs(PIDFCoefs pidfCoefs) {
...
+    default void setTuningSetpoint(double setpoint) {
+        setWithF(getControlMode(), setpoint, true);
     }
 
     double getClosedLoopTarget();
 
     default double getSetpoint() {
-        return isTuning() ? getTuningSetpoint() : getClosedLoopTarget();
+        boolean closedLoop = getControlMode() == ControlMode.Position || getControlMode() == ControlMode.Velocity;
+
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/pid/PIDFTalon.java:42]: blank line

File: src/main/java/frc/robot/utilities/pid/TrigonPIDController.java [hunk 17]
```diff
@@ -13,6 +14,21 @@ public TrigonPIDController(PIDCoefs pidCoefs) {
...
     }
 
+    @Override
+    public void setTuningSetpoint(double setpoint) {
+        setSetpoint(setpoint);
+    }
+
+    public void setSetpoint(double setpoint, boolean isTuning) {
+        if(isTuning() && !isTuning)
+            return;
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/pid/TrigonPIDController.java:24]: docs

File: src/main/java/frc/robot/utilities/pid/TrigonPIDController.java [hunk 18]
```diff
@@ -13,6 +14,21 @@ public TrigonPIDController(PIDCoefs pidCoefs) {
         isTuning = false;
     }
 
+    @Override
+    public void setTuningSetpoint(double setpoint) {
+        setSetpoint(setpoint);
+    }
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/pid/TrigonPIDController.java:20]: setTuningSetpoint is the same as setSetpoint......

**Review by Mboesn — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/RobotContainer.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (laviRZ) [reply src/main/java/frc/robot/RobotContainer.java:42]: Do you want to add a function that puts all of the sendables in the dashboard?

File: src/main/java/frc/robot/utilities/pid/PIDFMotor.java (hunk 9 above)
Comment (laviRZ) [reply src/main/java/frc/robot/utilities/pid/PIDFMotor.java:11]: PIDFMotor doesn't use a PIDFController, it uses the Talon's onboard PID

File: src/main/java/frc/robot/utilities/pid/PIDFMotor.java [hunk 19]
```diff
@@ -1,12 +1,22 @@
...
+
+    double get();
+
+    @Override
+    default void initSendable(SendableBuilder builder) {
+        PIDFConfigurable.super.initSendable(builder);
+        builder.setSmartDashboardType("PIDFMotor");
+        builder.setSafeState(() -> set(0));
+        builder.addDoubleProperty("output", this::get, isTuning() ? this::set : null);
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/pid/PIDFMotor.java:20]: with how the setpoint and isTuning is coded the setpoint will always override this

File: src/main/java/frc/robot/utilities/pid/PIDFMotor.java [hunk 20]
```diff
@@ -1,12 +1,22 @@
...
-    void tunePIDF(ControlMode controlMode);
+    void set(double output);
+
+    double get();
+
+    @Override
+    default void initSendable(SendableBuilder builder) {
+        PIDFConfigurable.super.initSendable(builder);
+        builder.setSmartDashboardType("PIDFMotor");
+        builder.setSafeState(() -> set(0));
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/pid/PIDFMotor.java:19]: what does this function do? I know it was used by CTRE just want you to check because it might work for the tuning

**Review by Mboesn — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/swerve/SwerveModule.java [hunk 21]
```diff
@@ -175,6 +181,24 @@ public SwerveModuleState getState() {
         return new SwerveModuleState(velocity, angle);
     }
 
+    public void setDesiredAngle(double angle, boolean isOpenLoop, boolean isTuning) {
```
[the remark this answers is not in the document]
Comment (laviRZ) [reply src/main/java/frc/robot/subsystems/swerve/SwerveModule.java:184]: look at where the value goes

File: src/main/java/frc/robot/subsystems/swerve/SwerveModule.java [hunk 22]
```diff
@@ -175,6 +181,24 @@ public SwerveModuleState getState() {
...
+        setDesiredState(
+                new SwerveModuleState(
+                        getLastDesiredState().speedMetersPerSecond,
+                        Rotation2d.fromDegrees(angle)),
+                isOpenLoop,
+                isTuning);
+    }
+
+    public void setDesiredSpeed(double speed, boolean isOpenLoop, boolean isTuning) {
```
[the remark this answers is not in the document]
Comment (laviRZ) [reply src/main/java/frc/robot/subsystems/swerve/SwerveModule.java:193]: look at where the value goes

File: src/main/java/frc/robot/utilities/pid/PIDConfigurable.java [hunk 23]
```diff
@@ -9,7 +9,11 @@
 
     double getSetpoint();
 
-    void setSetpoint(double setpoint);
+    void setSetpoint(double setpoint, boolean isTuning);
```
[the remark this answers is not in the document]
Comment (laviRZ) [reply src/main/java/frc/robot/utilities/pid/PIDConfigurable.java:12]: because the point of this `isTuning` is to check whether the caller of the function meant to change the setpoint while tuning in purpose, or it's a command, so we'll know to ignore the call.

File: src/main/java/frc/robot/utilities/pid/PIDFMotor.java (hunk 19 above)
Comment (laviRZ) [reply src/main/java/frc/robot/utilities/pid/PIDFMotor.java:20]: No, it sets the motor only after being changed.

File: src/main/java/frc/robot/utilities/pid/PIDFMotor.java (hunk 20 above)
Comment (laviRZ) [reply src/main/java/frc/robot/utilities/pid/PIDFMotor.java:19]: ask other programming team heads

File: src/main/java/frc/robot/utilities/pid/TrigonPIDController.java [hunk 24]
```diff
@@ -13,6 +14,18 @@ public TrigonPIDController(PIDCoefs pidCoefs) {
         isTuning = false;
     }
 
+    public void setSetpoint(double setpoint, boolean isTuning) {
```
[the remark this answers is not in the document]
Comment (laviRZ) [reply src/main/java/frc/robot/utilities/pid/TrigonPIDController.java:17]: Asked and answered, your honor.

File: src/main/java/frc/robot/subsystems/swerve/SwerveSS.java [hunk 25]
```diff
@@ -126,20 +130,24 @@ public void resetOdometry(Pose2d pose) {
...
     }
 
     /**
-     * Returns the current yaw (angle) of the robot
+     * Returns the current angle of the robot
      *
-     * @return the current yaw of the robot
+     * @return the current angle of the robot
      */
     public Rotation2d getAngle() {
```
Comment (Mboesn) [inline src/main/java/frc/robot/subsystems/swerve/SwerveSS.java:141]: state if CW or CCW

File: src/main/java/frc/robot/utilities/pid/TrigonPIDController.java [hunk 26]
```diff
@@ -13,6 +14,18 @@ public TrigonPIDController(PIDCoefs pidCoefs) {
...
+        // If the controller is tuning, and whatever called the function
+        // has nothing to do with tuning, we ignore the calling.
+        if(isTuning() && !isTuning)
+            return;
+        super.setSetpoint(setpoint);
+    }
+
+    public void setSetpoint(double setpoint) {
```
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/pid/TrigonPIDController.java:25]: overide

File: src/main/java/frc/robot/utilities/pid/TrigonPIDController.java (hunk 24 above)
Comment (Mboesn) [inline src/main/java/frc/robot/utilities/pid/TrigonPIDController.java:17]: overide

**Review by Mboesn — CHANGES_REQUESTED** [review]
